File: challenges/aoc/2023/2/1.py
# with open("sample.txt", 'r') as file:
with open("input.txt", 'r') as file:
    lines = file.readlines()

count = 0

for line in lines:
    id = 0

    for block in range(6, len(line)):
        if line[block] == ':':
            id = int(line[5:block])
            line = line[block+1:]
            break

    mainBlocks = line.split(';')

    isValid = True

    for b in mainBlocks:
        red, green, blue = 0, 0, 0
        subBlock = b.split(',')
        for s in subBlock:
            color = s.lstrip().split(' ')
            if color[1].startswith('r'):
                red += int(color[0])
            elif color[1].startswith('g'):
                green += int(color[0])
            elif color[1].startswith('b'):
                blue += int(color[0])

        if red > 12 or green > 13 or blue > 14:
            isValid = False

    if isValid:
        count += id

print("result: ", count)

File: challenges/aoc/2023/2/1.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <vector>

using namespace std;

int main() {
    // ifstream file("sample.txt");
    ifstream file("input.txt");

    string line;
    int count = 0;

    while (getline(file, line)) {
        int id = 0;

        for (int block = 6; block < line.length(); ++block) {
            if (line[block] == ':') {
                id = stoi(line.substr(5, block - 5));
                line = line.substr(block + 1);
                break;
            }
        }

        istringstream ss(line);
        vector<string> mainBlocks;

        while (getline(ss, line, ';')) {
            mainBlocks.push_back(line);
        }

        bool isValid = true;

        for (string& block : mainBlocks) {
            int red = 0, green = 0, blue = 0;
            istringstream subSS(block);
            string subBlock;

            while (getline(subSS, subBlock, ',')) {
                istringstream colorSS(subBlock);
                int value;
                string color;

                colorSS >> value >> color;

                if (color[0] == 'r') {
                    red += value;
                } else if (color[0] == 'g') {
                    green += value;
                } else if (color[0] == 'b') {
                    blue += value;
                }
            }

            if (red > 12 || green > 13 || blue > 14) {
                isValid = false;
                break;
            }
        }

        if (isValid) {
            count += id;
        }
    }

    file.close();

    cout << "Result: " << count << endl;

    return 0;
}

